Separate opam packages for lwt and async

bradlangel opened the pull request:
Based off of #3. Separates packages for lwt and async, so as to avoid depending on both. This follows the first four bullet points [outlined here](https://github.com/mfp/ocaml-statsd-client/pull/3#issuecomment-366538678). 
* Exposes three packages: `statsd-client`(with a deprecation warning for `Statsd_client.Lwt`), `statsd-client-lwt` and `statsd-client-async`.
* Packages can be installed locally by following these [opam instructions](https://opam.ocaml.org/doc/Packaging.html). 
* Gets rid of the compiler version dependency introduced in #2 for `statsd-client` and `statsd-client-lwt` but the dependency is kept for `statsd-client-async` since Async V0.10.0 depends on 4.0.1

bradlangel committed:
Moves Make functor to statsd_client_core.ml

* Adds statsd_client_core.ml and moves the `Make` functor to this file
* Exposes IO and T module signatures in statsd_client_core.mli

bradlangel committed:
Moves Statsd_client.Sync to Statsd_client_sync

Also module aliases Statsd_client_sync for backwards compatibility

bradlangel committed:
Moves Statsd_client.Lwt to Statsd_client_lwt

Includes module aliasing for Statsd_client_lwt in Statsd_client.lwt for
backwards compatibility.

bradlangel committed:
Moves Statsd_client.Async to Statsd_client_async

Includes module aliasing for Statsd_client_async as Statsd_client.async
for backwards compatibility.

bradlangel committed:
Separate packages for lwt, async and statsd-client

Moves lwt and async into their own directories in order to be able to
release independent packages. This approach is modeled off of what
ocaml-cohttp does. Furthermore, this also creates a statsd-client
directory in order to release a backwards compatible package.

bradlangel committed:
Removes async from backwards compatible package

bradlangel committed:
Removes compiler version from statsd-client pkg

bradlangel committed:
Adds deprecation warning for Statsd_client.Lwt

bradlangel committed:
Simplify interfaces by removing module T

bradlangel committed:
Statsd-client depends on statsd-client-lwt pkg

kevinqiu commented on statsd-client-async/statsd_client_core.ml:
```diff
@@ -0,0 +1 @@
+../statsd-client/statsd_client_core.ml
```
I'm not familiar with this syntax, what does this do? Pull in that file?

bradlangel replied:
symlinks to the file which async requires for building.

kevinqiu commented on statsd-client-async/statsd_client_async.mli:
```diff
@@ -0,0 +1,6 @@
+val gauge : ?sample_rate:float -> string -> int -> unit Async.Deferred.t
```
I wonder if there's way to avoid having to copy paste this interface between async + lwt. This seems to be the standard way of handling these separations though (looking at `ocaml-cohttp` for inspiration) so maybe not.

bradlangel replied:
yeah, there is. Check out my dogstatsd pull request. I modify the async and lwt signatures into one liners for these functions.

kevinqiu commented:
Looks reasonable to me. I'll let brendan take a look also.

bradlangel committed:
Makes Config module for ipaddr, port, etc.

Prior to this commit we were not exposing the two reference variables
ipaddr & port which resulted in a missing value error when sending a
metric. Furthermore, we expose the logger functions `log_debug` and
`log_error` as these were exposed before these separate package
commits.
In order to preserve backwards compatibility for the `statsd-client`
package with respect to these configuration variables, I had to symlink
to the `Statsd_client_lwt` module in the `statsd-client`
directory. Without doing this, setting `Statsd_client.{ipaddr,port}`
variables would work for `Statsd_client.Sync` functions but
`Statsd_client.Lwt` functions would throw an error that these variables
are undefined. By making this symlink in the `statsd-client` directory,
the `Statsd_client_lwt` module is built with the other modules and uses
the expected variables (as opposed to being a dependent package).
In the future, I'd like to get rid of this symlink behavior but this
requires a breaking change which at the moment we are avoiding for
the statsd-client package.
